Add transpose test cases

ErikSchierboom opened the pull request.

petertseng commented:
I like what I see for integers.

The exercise [description](https://github.com/exercism/x-common/blob/master/transpose.md) and [source](https://www.reddit.com/r/dailyprogrammer/comments/4msu2x/challenge_270_easy_transpose_the_input_text) specifiy text as well, how do we want to handle that?

There is a [PR on xRust](https://github.com/exercism/xrust/pull/151) that assume text inputs, and the additional implication here is that the rows/columns need not all be equal sizes. Some additional cases to deal with those would probably be appropriate.

ErikSchierboom commented:
Good idea, I'll add some string cases too that take into account varying row/column lengths.

ErikSchierboom commented:
@petertseng Having read the exercise, it does indeed explicitly mention text. If we would also like to support transposing non-string arrays, I think it makes sense to have a multiline string passed as an array of strings (each line being an array entry). What do you think?

Insti commented:
Is it likely that a solution that solves for arrays of numbers will not also solve arrays of strings?

ErikSchierboom commented:
@Insti Yes, but it is not clear from the exercise text that the string is passed as an array of strings instead of a string with newline characters.

Insti commented:
The description: 

> Write a program that takes input text and outputs it -- transposed.

Would seem to suggest that the input (and output) is text with newlines which would make the arrays in the .json file inappropriate?

ErikSchierboom commented:
@petertseng @Insti Maybe we should then only focus on text and not on integers?

Insti commented:
Yes, probably.

But it looks like fsharp is the only track that implements this problem?
- In which case just change the description to match what you want to do :smiley:

ErikSchierboom commented:
@Insti Well, I was coming around to that :) The only thing that would be modified is that multiple lines are passed as an array of strings.

Insti commented:
Opinion: 
A problem that involves transforming text into text is more interesting than one that turns an x by y array into a y by x array.
(Even if it's essentially the same thing with added split/join)

ErikSchierboom commented:
@Insti Agreed. I'll close this PR.

Insti commented:
A [Word square](https://en.wikipedia.org/wiki/Word_square) would make an interesting test case.
```
H E A R T
E M B E R
A B U S E
R E S I N
T R E N D
```

Insti commented:
And [Word rectangles (from the same page)](https://en.wikipedia.org/wiki/Word_square#Word_rectangles)
```
F R A C T U R E
O U T L I N E D
B L O O M I N G
S E P T E T T E
```

ErikSchierboom commented:
@Insti On second thought, maybe I should update the PR with some new, string-based examples?

Insti commented:
@ErikSchierboom I think that's a good idea.

ErikSchierboom commented:
@Insti @petertseng I've updated the test cases to work with strings. Better?

Insti commented on transpose.json:
```diff
@@ -0,0 +1,186 @@
+      "JSON doesn't allow for multi-line strings, so all multi-line input is ",
+      "presented here as arrays of strings. It's up to the test generator to join the ", 
+      "lines together with line breaks."
+  ],
```
```
{"multi_line_string":"Multi\nLine\nString\n"}
```

Insti replied:
Perhaps it should be: `JSON doesn't allow for elegant formatting of multi-line strings, so all multi-line input and output is`...

ErikSchierboom replied:
That seems like a good wording. I copied the text from the existing JSON data from [food-chain](https://github.com/exercism/x-common/blob/master/food-chain.json) by the way.

Insti commented:
This new version looks great :+1:

It would be good with some simple initial test cases.
```
A1

A
1
```
and 
```
ABC
123 

A1
B2
C3
```
and their inverses.

ErikSchierboom commented:
Good idea. Will add them.

ErikSchierboom commented:
@Insti done!

petertseng commented on transpose.json:
```diff
@@ -0,0 +1,206 @@
+          "ni",
+          "en",
+          ".e",
+          "." 
```
in the original text, the period belongs to the second line, so there should be a space before it in the transposed text

petertseng commented on transpose.json:
```diff
@@ -0,0 +1,206 @@
+        ],
+      "expected": [ 
+          "TEASER",
+          "EASER",
```
in the original text, the E's were all in the same row, so in the transposed text they should all be in the same column

ErikSchierboom replied:
I've fixed this in the updated PR.

Insti commented:
There doesn't seem to be a test case where the first line is longer than the second line.

ErikSchierboom commented:
@Insti @petertseng I've updated the incorrect test cases, and added a test case where the first line is longer than the second line.

I've also looked at the [xRust PR](https://github.com/exercism/xrust/pull/151) and it appears their test case also had some problems, where input spaces where stripped in the output.

petertseng commented:
I think everything is in order (I haven't verified with code though).

It might be worth adding to the README a few things, because otherwise the problem seems unspecified.

* What happens when line are different lengths (pad to the left with spaces, but don't pad to the right)
* If a column in the input text contains only spaces on its bottom-most rows, should the corresponding output row contain the spaces in its right-most columns?
  * The [source](https://www.reddit.com/r/dailyprogrammer/comments/4msu2x/challenge_270_easy_transpose_the_input_text) says they shouldn't.
  * I can see argument for why it should because that's what's in the input text, and we're just transposing it! And that's what the test cases expect
  * I absolutely do not care what we choose, and we currently chose "keep the spaces", so we can roll with that.

ErikSchierboom commented:
@petertseng The README has been updated as suggested. Is it clear enough?

petertseng commented on transpose.md:
```diff
@@ -16,3 +16,44 @@ CF
+
+And transposing:
+
+Therefore, transposing this matrix:
```
remove this line, I think

petertseng commented on transpose.md:
````diff
@@ -16,3 +16,44 @@ CF
+
+Therefore, transposing this matrix:
+
+ ```
````
extra space before the backticks (but I checked and it still gets rendered)

petertseng commented:
I like what I see here.

👍  from me.

ErikSchierboom commented:
@petertseng I've fixed the remaining issues and I'll merge it. Thanks for the help!

petertseng commented on transpose.json:
```diff
@@ -0,0 +1,231 @@
+          "          o  n",
+          "          v  d",
+          "          e  .",
+          "          ,  "
```
I think I missed a small thing - no spaces after the comma, right?

(I also haven't checked with code, so I may have missed others, but I  least see this)

petertseng replied:
I'll send the PR. This was the only thing.
